Hw03 Success

Minor152 opened the pull request:
## Домашнее задание №3 «Шахматная доска»

### Чек-лист студента ([Что это?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0))
- [x] Я перечитал задание после решения. [Зачем перечитывать задание?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0#user-content-%D0%97%D0%B0%D1%87%D0%B5%D0%BC-%D0%BF%D0%B5%D1%80%D0%B5%D1%87%D0%B8%D1%82%D1%8B%D0%B2%D0%B0%D1%82%D1%8C-%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5)
- [x] Я запустил `go mod tidy`.
- [x] Я сравнил объем кода с объемом авторского решения. [Зачем сверять объем кода?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0#user-content-%D0%97%D0%B0%D1%87%D0%B5%D0%BC-%D1%81%D0%B2%D0%B5%D1%80%D1%8F%D1%82%D1%8C-%D0%BE%D0%B1%D1%8A%D0%B5%D0%BC-%D0%BA%D0%BE%D0%B4%D0%B0)
  - main.go ~25 строк

### Критерии оценки
- [ ] Пайплайн зелёный - 4 балла
- [ ] Понятность и чистота кода - до 2 баллов

#### Зачёт от 4 баллов

Commit:
Hw03 Success

Commit:
fix

Commit:
fix

hungry-honeybadger commented on hw03_chessboard/init.go:
```diff
@@ -0,0 +1,32 @@
+	CHESSBOARD strings.Builder
+)
+
+func init() {
```
Использование метода init считается плохой практикой в go, хотя в раде случаев и допускается. Но загружать в или просить пользователя в нем  что либо вводить абсолютно не допустимо. Вот представим  что данный код решили решили использовать для того, что бы встроить в другой продукт. например в веб сервер. И поведение когда веб сервер при запуске просит ввести какие либо параметры . И для разработчика веб сервера это будет не явное поведение. 2 момент. Мы тут пытаемся вывести какую то-информацию в консоль, но мы уверены что мы можем ее вывести? Что у нас вообще есть консоль?

RuChik152 replied:
Добрый день.

Про возврат функции понял. Исправлю.

с остальным нет. 

Про init нам на лекции лектор сказал что init использовать можно и нет в этом проблемы


Так же у нас в описании к ДЗ как подсказка, что нужно использовать 
`fmt.Scanf("%d", &size)`
и исходя из описания задания, я посчитал, что предпологается использовать код как некую CLI утилиту, которая запуститься, примет от пользователя данные, вернет в консоль результат и завершит работу.

Если нет то немного не понятна логика зачем указывать как подсказку 
`fmt.Scanf("%d", &size)`

Minor152 replied:
Добрый день.

Про возврат функции понял. Исправлю.

с остальным нет.

Про init нам на лекции лектор сказал что init использовать можно и нет в этом проблемы

Так же у нас в описании к ДЗ как подсказка, что нужно использовать
fmt.Scanf("%d", &size)
и исходя из описания задания, я посчитал, что предпологается использовать код как некую CLI утилиту, которая запуститься, примет от пользователя данные, вернет в консоль результат и завершит работу.

Если это не так, то немного не понятна логика зачем тогда указывать как подсказку
fmt.Scanf("%d", &size)
и есть небольшой диссонанс по ряду тем с тем что говорили на лекциях и с тем что указано в замечании.

hungry-honeybadger replied:
`fmt.Scanf("%d", &size)` должен быть в методе main. В init не должно быть взаимодействия с пользователем

hungry-honeybadger commented on hw03_chessboard/init.go:
```diff
@@ -0,0 +1,32 @@
+	"strings"
+)
+
+var (
```
В данном случае нет никакого смысла использовать переменные уровни пакета. Они лишь делают наш код непригодным к использованию в многопоточным режиме. Рекомендую стараться использовать переменные с уровнем видимости на уровне 1 функции

Minor152 replied:
исправил

hungry-honeybadger commented on hw03_chessboard/main.go:
```diff
@@ -1,5 +1,17 @@
+	for l := 0; l < SizeRow; l++ {
+		for c := 0; c < SizeColumn; c++ {
+			if c == 0 {
+				fmt.Fprintf(&CHESSBOARD, "\n")
```
Для записи в strings.Builder лучше все таки использовать его методы WriteRune и WriteString

Minor152 replied:
исправил

hungry-honeybadger reviewed (changes requested):
Результат работы программы не является шахматной доской 
![image](https://github.com/user-attachments/assets/a869df11-ef71-45bf-940e-4484c3356f54)

Commit:
Изменил main.go
Убрал использование кастомной функции init
Исправил замечания которые были указаны при прпроверке
Использовал структуру.

Commit:
fix go lint

Minor152 committed:
fix

hungry-honeybadger commented on hw03_chessboard/board.go:
```diff
@@ -0,0 +1,59 @@
+	fmt.Println("Enter the number of rows: ")
+	_, err = fmt.Scanf("%d", &b.rows)
+	if err != nil {
+		log.Fatalln(fmt.Errorf("%w", err))
```
Нет никакого смысла от `fmt.Errorf("%w", err)`. Можно было сразу вывести err. `fmt.Errorf("%w", err)` нужен если мы хотим как то расширить инфу об ошибке

hungry-honeybadger commented on hw03_chessboard/board.go:
```diff
@@ -0,0 +1,59 @@
+		if l%2 != 0 {
+			b.board.WriteString("\n")
+			for c := 0; c < b.columns; c++ {
+				if c%2 == 0 {
```
Два if тут лишнее. Бес особых правок это меняется if else. Так результат операции `l%2` это либо 0 либо не 0  а это как раз if else. Либо можно посмотреть на вариант `(l+c)%2`
